### PR TITLE
Correction for variations of fiber flat-field with humidity

### PR DESCRIPTION
**Pull request by julienguy**
This PR addresses the issue of the fiber flat-field variations with humidity.

- add io (in `desispec/io/fiberflat_vs_humidity.py`)
  - `get_humidity(night,expid,camera)` function to retrieve the telemetry humidity for one night/expid/camera.
  - `read_fiberflat_vs_humidity` function to read the new fiberflat_vs_humidity calibration dataset
  - add `findfile("fiberflatexp")` : there is now one fiberflat frame per science frame

- add script to compute median flatfields per camera/fiber in bins of humidity
  - file: `bin/desi_compute_fiberflat_vs_humidity`
  - scans all daily/calibnight flatfield from 2021
  - group them in bins of humidity (bin edges are chosen automatically with the constrain of 10 flats per bin)
  - median in the bin
  - save wavelength, median fiberflat for each bin, in fits file that one can read with `read_fiberflat_vs_humidity`

Example for b1, file `/global/cfs/cdirs/desi/users/jguy/teststand/desi_spectro_calib/spec/sm10/fiberflat-vs-humidity-sm10-b1.fits`
Figure_1
Figure_1_zoom
variation-from-cross-correlation

- add function `compute_humidity_corrected_fiberflat` in `desispec/fiberflat_vs_humidity.py`
- add script `bin/desi_compute_humidity_corrected_fiberflat`
  - retrieves  telemetry humidity for dome flat and for the current exposure
  - **fit** the humidity for the current dome flat and for the **sky fibers of the current exposure**
  -  returns  corrected flat = dome_flat / flat_vs_humidity_model(fit of humidity from dome flat) * flat_vs_humidity_model(fit of humidity from sky fibers in current exposure) 

 ```
usage: desi_compute_humidity_corrected_fiberflat [-h] -i INFILE --fiberflat FIBERFLAT -o OUTFILE

Compute a fiberflat corrected for variations with humidity.

optional arguments:
  -h, --help            show this help message and exit
  -i INFILE, --infile INFILE
                        path of DESI exposure frame fits file
  --fiberflat FIBERFLAT
                        path of DESI fiberflat fits file
  -o OUTFILE, --outfile OUTFILE
                        path of output fiberflar file
  ```
- add the logic in `bin/desi_proc`  to compute and use a fiberflat per exposure instead of the nightly fiberflat.

- I computed and stored in my checkout of the calibration the humidity dependent flatfield files. They
are found by the code with  the keyword FIBERFLATVSHUMIDITY in the calib yaml file. I have computed them only
for the blue cameras.

- In order to test the code, use at NERSC: `export DESI_SPECTRO_CALIB=/global/cfs/cdirs/desi/users/jguy/teststand/desi_spectro_calib`

Example (close to worst case scenario: bright sky and very large variation of humidity during the night)
```
desi_proc -n 20211215 -e 00114218 --batch
```
Looking at b1 (similar plots for all blue cameras):
```
ds9 -scale LIMITS -200 200 /global/cfs/cdirs/desi/spectro/redux/{daily,jguy}/exposures/20211215/00114218/sframe-b1-00114218.fits
```
Screenshot from 2022-01-03 16-45-35

```
plot_frame -i /global/cfs/cdirs/desi/spectro/redux/{daily,jguy}/exposures/20211215/00114218/sframe-b1-00114218.fits --fiber 251
```
Screenshot from 2022-01-03 16-48-42

Humidities from telemetry and fit are saved in header of exposure fiberflat. We are able to predict the humidity in the shack with a precision of ~1%.

```
desi_header -i /global/cfs/cdirs/desi/spectro/redux/jguy/exposures/20211215/00114218/fiberflatexp-b1-00114218.fits 

...
EXPTHUM =                11.43 / exposure humidity from telemetry               
EXPFHUM =    12.24315219834253 / exposure humidity from flat fit                
CALFHUM =    29.42873837268855 / dome flat humidity from flat fit               
CALTHUM =    29.07769230769231 / dome flat humidity from telemetry       
```

### CODE REVIEW

**Comment by julienguy** [comment]
Again, tests are failing because of an independent issue (fitsio incompatibility in the test setup).

**Comment by moustakas** [comment]
Wowza, nice!

**Comment by sbailey** [comment]
I am running a 3-night test production using this branch in `/global/cfs/cdirs/desi/users/sjbailey/spectro/redux/humidity`, including two nights with known variable humidity 20211215 and 20211220 and 20210515 with a mix of dark and bright tiles.  At the rate the jobs are running now they are likely to finish by morning.  Redshifts are "pernight" which are directly comparable to previous prods (unlike "cumulative" which span nights).  These re-use previously run calibs and only reprocess science exposures.

Notes to self:
```
godesi test-master && module swap desispec/sjb2
export DESI_SPECTRO_CALIB=/global/cfs/cdirs/desi/users/jguy/teststand/desi_spectro_calib

myprod humidity
cdprod
copyprod $rx/daily . --night 20211215
copyprod $rx/daily . --night 20211220
copyprod $rx/f4 . --night 20210515
rm -r exposures
rm -r preproc

desi_run_night --proc-obstypes science --z-submit-types pernight --system-name cori-knl -q regular -n 20211215
desi_run_night --proc-obstypes science --z-submit-types pernight --system-name cori-knl -q regular -n 20210515
desi_run_night --proc-obstypes science --z-submit-types pernight --system-name cori-knl -q regular -n 20211220
```

File: py/desispec/scripts/humidity_corrected_fiberflat.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        log.info("No information on fiberflat vs humidity for camera {}, simply link the input fiberflat".format(frame_header["CAMERA"]))
+        if not os.path.islink(args.outfile) :
+            os.symlink(args.fiberflat,args.outfile)
+        return 0
+
+    # read fiberflat
+    calib_fiberflat = read_fiberflat(args.fiberflat)
+
+    # read mean fiberflat vs humidity
+    filename = cfinder.findfile("FIBERFLATVSHUMIDITY")
```
Comment (sbailey) [inline py/desispec/scripts/humidity_corrected_fiberflat.py:59]: in $DESI_SPECTRO_CALIB, let's include a date in the name of the fiberflat-vs-humidity-*.fits files to establish a naming scheme that naturally supports having having multiple versions.  e.g. in the future we'll need a different correction file after the collimators are replaced.  I know the yaml structure allows the names to be arbitrary, but let's not have the first one be the one without a date.

File: bin/desi_compute_fiberflat_vs_humidity [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+    humidity_table_filename=f"humidity_table_{cam}.csv"
+    if not os.path.isfile(humidity_table_filename) :
+
+        vals=dict()
+        vals["NIGHT"]=[]
+        vals["HUMIDITY"]=[]
+
+        # loop on nights with fiberflat
+        for flat_filename in sorted(glob.glob(f"/global/cfs/cdirs/desi/spectro/redux/daily/calibnight/{year}*/fiberflatnight-{cam}-*.fits")) :
```
Comment (sbailey) [inline bin/desi_compute_fiberflat_vs_humidity:38]: Thanks for including this script to document how the files are made and give us a starting point for making updates.  We could deal with this later, but eventually we'll need to change the hardcoded `year` to options for a date range (e.g. to make a model after replacing the collimators), and update the hardcoded production path to use `desispec.io.specprod_dir()` (e.g. if we wanted to make an updated cleaner model using a consistent reference production that had used better CCD calibrations).

File: bin/desi_compute_fiberflat_vs_humidity [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+                #print(desi_filename)
+                spectcons=fitsio.read(desi_filename,"SPECTCONS")
+                ii=(spectcons["unit"]==spectro)
+                if np.sum(ii)==0 :
+                    log.warning(f"no spectro {spectro} in {desi_filename}")
+                    continue
+                bhumid=float(spectcons["BHUMID"][ii])
+                if bhumid<=0 :
+                    log.warning(f"unphysical humidity value = {bhumid}")
+                    continue
```
Comment (sbailey) [inline bin/desi_compute_fiberflat_vs_humidity:63]: Minor: Let's also be robust to bhumid > 100 and missing BHUMID column.

File: py/desispec/io/fiberflat_vs_humidity.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    log=get_logger()
+    raw_filename=findfile("raw",night=night,expid=expid)
+    table=fitsio.read(raw_filename,"SPECTCONS")
+    keyword="{}HUMID".format(camera[0].upper())
+    unit=int(camera[1])
+    selection=(table["unit"]==unit)
+    if np.sum(selection)==0 :
+        log.warning("no unit '{}' in '{}'".format(unit,raw_filename))
+        return np.nan
+    humidity=float(table[keyword][selection][0])
```
Comment (sbailey) [inline py/desispec/io/fiberflat_vs_humidity.py:18]: Consider checking for physical values here, and being robust to the column completely missing (same behavior as the camera missing from the table).

File: py/desispec/fiberflat_vs_humidity.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    for loop in range(2) :
+        # remove mean variation from fiber to fiber
+        med   = np.median(tmp_flat,axis=-1)
+        tmp_flat /= med[:,:,None]
+        # remove average over fibers
+        for index in range(tmp_flat.shape[0]) :
+            med =  np.median(tmp_flat[index],axis=0)
+            tmp_flat[index] /= med[None,:]
+
+    chi2 = np.sum(np.sum(tmp_ivar*(tmp_flux-tmp_flat)**2,axis=-1),axis=-1)
```
Comment (sbailey) [inline py/desispec/fiberflat_vs_humidity.py:64]: I think I know what you are doing here, but please add some comments for our future selves about what this fit is doing and why.

File: py/desispec/fiberflat_vs_humidity.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    if flat1 is None :
+        log.info("use input humidity = {:.2f}".format(current_humidity))
+        flat1  = _interpolated_fiberflat_vs_humidity(mean_fiberflat_vs_humidity , humidity_array, current_humidity)
+
+    # apply humidity correction to current calib fiberflat
+    fiberflat = copy.deepcopy(calib_fiberflat)
+    fiberflat.fiberflat = calib_fiberflat.fiberflat/flat2*flat1
+    fiberflat.header["EXPTHUM"] = (current_humidity,"exposure humidity from telemetry")
+    fiberflat.header["EXPFHUM"] = (hum1,"exposure humidity from flat fit")
+    fiberflat.header["CALFHUM"] = (hum2,"dome flat humidity from flat fit")
```
Comment (sbailey) [inline py/desispec/fiberflat_vs_humidity.py:126]: Let's print warnings or errors if the fitted humidity is very different from the telemetry humidity.

**Review by sbailey — CHANGES_REQUESTED** [review]
Algorithmically this looks *great*.  The 3 test nights have significantly improved sframesky residuals and reduced pileup of redshifts, e.g. the most dramatic case of 20211215:
image
FTR: that isn't exactly an apples-to-apples comparison since daily only has cumulative redshifts, while the test prod only has single nights, and this plot mixes tiles that were observed across multiple nights in there giving daily and advantage (and still the test prod is better!)

I put a number of inline comments about maintainability, docstrings, future calibs, etc.  I think the most critical to address is about not having absolute symlinks in the production, since that doesn't work well with archiving prods elsewhere.

**Comment by sbailey** [comment]
A slight algorithmic concern that I haven't tracked down: the number of ZWARN != 0 targets went up by 1% on 20211215 even after filtering to just tiles observed only on that night.  That might be ok (even correct!) but noting it here in case someone else can help cross check those.

On 20211220 the number of ZWARNs was much closer.

File: py/desispec/io/fiberflat_vs_humidity.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        wave is 1D [nwave] (and in Angstrom)
+        header (fits header)
+    """
+
+    with fits.open(filename, uint=True, memmap=False) as fx:
+        header = fx[0].header
+        wave  = native_endian(fx["WAVELENGTH"].data.astype('f8'))
+        fiberflat = list()
+        humidity  = list()
+        for index in range(20) :
```
[the remark this answers is not in the document]
Comment (julienguy) [reply py/desispec/io/fiberflat_vs_humidity.py:40]: done

File: py/desispec/fiberflat_vs_humidity.py (hunk 6 above)
Comment (julienguy) [reply py/desispec/fiberflat_vs_humidity.py:126]: done

File: bin/desi_compute_fiberflat_vs_humidity (hunk 3 above)
Comment (julienguy) [reply bin/desi_compute_fiberflat_vs_humidity:66]: done

File: py/desispec/scripts/humidity_corrected_fiberflat.py (hunk 1 above)
Comment (julienguy) [reply py/desispec/scripts/humidity_corrected_fiberflat.py:59]: done

File: py/desispec/io/fiberflat_vs_humidity.py (hunk 4 above)
Comment (julienguy) [reply py/desispec/io/fiberflat_vs_humidity.py:18]: done

File: py/desispec/fiberflat_vs_humidity.py (hunk 5 above)
Comment (julienguy) [reply py/desispec/fiberflat_vs_humidity.py:92]: done

File: py/desispec/fiberflat_vs_humidity.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+
+import numpy as np
+import copy
+
+from desiutil.log import get_logger
+from desispec.fiberflat import apply_fiberflat
+from desispec.fiberbitmasking import get_skysub_fiberbitmask_val
+
+def _interpolated_fiberflat_vs_humidity(fiberflat_vs_humidity , humidity_array, humidity_point) :
```
[the remark this answers is not in the document]
Comment (julienguy) [reply py/desispec/fiberflat_vs_humidity.py:17]: done

File: py/desispec/scripts/humidity_corrected_fiberflat.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        # need full frame to adjust correction on data
+        frame = read_frame(args.infile)
+    else :
+        frame = None
+
+    cfinder = CalibFinder([frame_header])
+    if not cfinder.haskey("FIBERFLATVSHUMIDITY"):
+        log.info("No information on fiberflat vs humidity for camera {}, simply link the input fiberflat".format(frame_header["CAMERA"]))
+        if not os.path.islink(args.outfile) :
+            os.symlink(args.fiberflat,args.outfile)
```
[the remark this answers is not in the document]
Comment (julienguy) [reply py/desispec/scripts/humidity_corrected_fiberflat.py:52]: For today, I have implemented the symbolic links. We can consider in another PR the 'calibnight/YEARMMDD/fiberflatdefault-.fits' idea.

File: bin/desi_compute_fiberflat_vs_humidity (hunk 2 above)
Comment (julienguy) [reply bin/desi_compute_fiberflat_vs_humidity]: done

**Review by sbailey — APPROVED** [review]

**Pull request merged**